scipy/stats/_mannwhitneyu.py

```diff
@@ -0,0 +1,430 @@
+        shape_new = np.array((m+1, n+1, k+1))
+        if np.any(shape_new > shape_old):
+            shape = np.maximum(shape_old, shape_new)
+            fmnks = -np.ones(shape)             # create the new array
+            m0, n0, k0 = shape_old
+            fmnks[:m0, :n0, :k0] = self._fmnks  # copy remembered values
+            self._fmnks = fmnks
+
+    def _f(self, m, n, k):
+        '''Recursive implementation of function of [3] Theorem 2.5'''
+
+        fmnks = self._fmnks  # for convenience
+
+        # [3] Theorem 2.5 Line 1
+        if k < 0 or m < 0 or n < 0 or k > m*n:
+            return 0
+
+        # if already calculated, return the value
+        if fmnks[m, n, k] >= 0:
+            return fmnks[m, n, k]
+
+        if k == 0 and m >= 0 and n >= 0:  # [3] Theorem 2.5 Line 2
+            fmnk = 1
+        else:   # [3] Theorem 2.5 Line 3 / Equation 3
+            fmnk = self._f(m-1, n, k-n) + self._f(m, n-1, k)
+
+        fmnks[m, n, k] = fmnk  # remember result
+
+        return fmnk
+
+
+# Maintain state for faster repeat calls to mannwhitneyu w/ method='exact'
+_mwu_state = _MWU()
+
+
+def _tie_term(ranks):
+    """Tie correction term"""
+    # element i of t is the number of elements sharing rank i
+    _, t = np.unique(ranks, return_counts=True, axis=-1)
+    return (t**3 - t).sum(axis=-1)
+
+
+def _get_mwu_z(U, n1, n2, ranks, axis=0, continuity=True):
+    '''Standardized MWU statistic'''
+    # Follows mannwhitneyu [2]
+    mu = n1 * n2 / 2
+    n = n1 + n2
+
+    # Tie correction according to [2]
+    tie_term = np.apply_along_axis(_tie_term, -1, ranks)
+    s = np.sqrt(n1*n2/12 * ((n + 1) - tie_term/(n*(n-1))))
+
+    # equivalent to using scipy.stats.tiecorrect
+    # T = np.apply_along_axis(stats.tiecorrect, -1, ranks)
+    # s = np.sqrt(T * n1 * n2 * (n1+n2+1) / 12.0)
+
+    numerator = U - mu
+
+    # Continuity correction.
+    # Because SF is always used to calculate the p-value, we can always
+    # _subtract_ 0.5 for the continuity correction. This always increases the
+    # p-value to account for the rest of the probability mass _at_ q = U.
+    if continuity:
+        numerator -= 0.5
+
+    # no problem evaluating the norm SF at an infinity
+    with np.errstate(divide='ignore', invalid='ignore'):
+        z = numerator / s
+    return z
+
+
+def _mwu_input_validation(x, y, use_continuity, alternative, axis, method):
+    ''' Input validation and standardization for mannwhitneyu '''
+    # Would use np.asarray_chkfinite, but infs are OK
+    x, y = np.atleast_1d(x), np.atleast_1d(y)
+    if np.isnan(x).any() or np.isnan(y).any():
+        raise ValueError('`x` and `y` must not contain NaNs.')
+    if np.size(x) == 0 or np.size(y) == 0:
+        raise ValueError('`x` and `y` must be of nonzero size.')
+
+    bools = {True, False}
+    if use_continuity not in bools:
+        raise ValueError(f'`use_continuity` must be one of {bools}.')
+
+    alternatives = {"two-sided", "less", "greater"}
```

Comment:
Is there an advantage to write {} instead of []?

Reply:
Practically speaking, not for a set this small. But a `set` is the more appropriate data structure here since we're  only interested in set membership, not order.

Reply:
agree